server.py

```diff
@@ -131,23 +132,29 @@ def assign_robot_to_client(self, ws):
         message = message.json()
         # Set robot's attributes according to data in message
         # While selection is not confirmed, it is still possible to choose cards
-        # TODO: not only by pressing key but also when time's up
-        if not message["interface_data"]["confirmed"]:
-            # TODO: this part only sets the POWER DOWN attribute,
-            # it doesn't affect anything else.
-            robot.power_down = message["interface_data"]["power_down"]
+        # TODO: this part only sets the POWER DOWN attribute,
+        # it doesn't affect anything else.
+        robot.power_down = message["interface_data"]["power_down"]
 
+        if not message["interface_data"]["confirmed"]:
             # Set robot's program with chosen cards
             selection = message["interface_data"]["my_program"]
             for card_index in selection:
                 if card_index is not None:
                     robot.program[selection.index(card_index)] = robot.dealt_cards[card_index]
-
         # choice of cards was blocked by the player
         else:
-            # Add the rest of the cards to used cards pack
             robot.selection_confirmed = True
-            await self.play_round()
+            # Check if all robots selected cards.
+            all_selection_confirmed = self.state.all_selection_confirmed()
+            if not all_selection_confirmed:
+                selection_number = self.state.count_selection_number()
+                # If last robot doesnt selected his cards, the timer starts.
+                if selection_number == len(self.state.robots) - 1:
+                    await self.send_message("timer_start")
+                    asyncio.create_task(self.timer(self.state.game_round))
```

Comment:
Muze se vubec stat, ze se podminka v tak zavolane funkci nesplni? 

Dal v metode timer jsou radky:
```
 if game_round == **self.state.game_round**:
     self.choose_random_card()
     await self.apply_effects()
```

Znamena to, ze volani te funkce vzdy bude vyhodnoceno jako True?

Reply:
všechny ty podmínky navíc které asi nechápeš jsou tak ošetřené kvůli tomu, když jelo více časovačů zároveň, už z dalšího kola. Protože pokud by game_round nesedělo, neudělá se nic. Představ si, že zahraješ jedno kolo, oba dva potvrdí karty ještě než skončí časovač. časovač 1 stále běží. Mezitím dokončíš jedeš další kolo, běží druhý časovač. A bum, skončí první časovač a karty  které máš zatím vybrané ti zmizí, provedou se efekty a rozdají se karty nové a jsi v dalším kole. Strašně těžko se to vysvětluje, ale zabralo mi to hodně času, než jsem vychytala všechny mouchy.
Takže pokud to bude vyhodnoceno jako false, tak se nestane nic, ale to si myslím že tam není nutné psát, alespoň to nemělo žádný efekt, když jsem to tam dala.

Stejně tak jsem bojovala se selection_number, kvůli dvěma časovačům. Běží první časovač, oba stihnout potvrdit, takže selection_number = 2. Další kolo, jeden hráč potvrdí, jede druhý časovač, selection_number = 1, a když doběhne první časovač, zjistí že je to jedna a provedlo by se to, takže ještě se kontroluje to kolo. Takže ta podmínka se nemusí splnit. Je to fakt složité.